File: sdk/java/dagger-codegen-maven-plugin/src/main/java/io/dagger/codegen/introspection/Directive.java
package io.dagger.codegen.introspection;

import java.util.List;

public class Directive {

  private String name;
  private List<DirectiveArg> args;

  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public List<DirectiveArg> getArgs() {
    return args;
  }

  public void setArgs(List<DirectiveArg> args) {
    this.args = args;
  }

  /**
   * Get the value of a directive argument by name.
   *
   * @return the raw JSON string value, or null if not found
   */
  public String getArgValue(String argName) {
    if (args == null) {
      return null;
    }
    for (DirectiveArg arg : args) {
      if (argName.equals(arg.getName())) {
        return arg.getValue();
      }
    }
    return null;
  }

  /**
   * Get the @expectedType name from a list of directives, if present. Returns the unquoted type
   * name from @expectedType(name: "Foo").
   */
  public static String getExpectedType(List<Directive> directives) {
    if (directives == null) {
      return null;
    }
    for (Directive d : directives) {
      if ("expectedType".equals(d.getName())) {
        String val = d.getArgValue("name");
        if (val != null) {
          // The value comes as a JSON-encoded string, e.g. "\"Container\""
          // Strip surrounding quotes if present
          if (val.startsWith("\"") && val.endsWith("\"")) {
            val = val.substring(1, val.length() - 1);
          }
          return val;
        }
      }
    }
    return null;
  }

  @Override
  public String toString() {
    return "Directive{" + "name='" + name + '\'' + ", args=" + args + '}';
  }
}

File: sdk/dotnet/sdk/Dagger.SDK.SourceGenerator/Types/Directive.cs
using System.Text.Json.Serialization;

namespace Dagger.SDK.SourceGenerator.Types;

public class DirectiveArg
{
    [JsonPropertyName("name")]
    public required string Name { get; set; }

    [JsonPropertyName("value")]
    public required string Value { get; set; }
}

public class Directive
{
    [JsonPropertyName("name")]
    public required string Name { get; set; }

    [JsonPropertyName("args")]
    public DirectiveArg[] Args { get; set; } = [];

    /// <summary>
    /// If this is an @expectedType directive, return the type name; otherwise null.
    /// </summary>
    public string? GetExpectedType()
    {
        if (Name != "expectedType") return null;
        foreach (var arg in Args)
        {
            if (arg.Name == "name")
            {
                // Value comes as "\"Container\"" — strip surrounding quotes
                return arg.Value.Trim('"');
            }
        }
        return null;
    }
}
